Don't join master nodes or accept join requests of old and too new nodes

martijnvg opened the pull request:
If the version of a node is lower than the minimum supported version or higher than the maximum hypothetical supported version, a node shouldn't be allowed to join and nodes should join that elected master node

bleskes commented on core/src/test/java/org/elasticsearch/discovery/zen/ZenDiscoveryTests.java:
```diff
@@ -215,4 +216,88 @@ public void handleException(TransportException exp) {
+        Settings nodeSettings = Settings.settingsBuilder()
+                .put("discovery.type", "zen") // <-- To override the local setting if set externally
+                .build();
+        String nodeName = internalCluster().startNode(nodeSettings, Version.V_1_6_0);
```
I wonder how long this would work ;)

bleskes commented on core/src/test/java/org/elasticsearch/discovery/zen/ZenDiscoveryTests.java:
```diff
@@ -215,4 +216,88 @@ public void handleException(TransportException exp) {
+        ZenDiscovery zenDiscovery = (ZenDiscovery) internalCluster().getInstance(Discovery.class, nodeName);
+
+        DiscoveryNode node = new DiscoveryNode("_node_id", new LocalTransportAddress("_id"), Version.V_2_0_0);
+        final IllegalStateException[] e = new IllegalStateException[1];
```
can we use atomic reference?

martijnvg replied:
the reason I used an array here, is because setting/getting happens on the same thread. but I can change it to atomic reference.

bleskes commented on core/src/main/java/org/elasticsearch/discovery/zen/ZenDiscovery.java:
```diff
@@ -453,7 +456,15 @@ public void onFailure(String source, @Nullable Throwable t) {
-    private boolean joinElectedMaster(DiscoveryNode masterNode) {
+    boolean joinElectedMaster(DiscoveryNode masterNode) {
+        // Sanity checks: likely we don't end up here, because serialization is likely to fail.
+        if (masterNode.getVersion().before(version.minimumCompatibilityVersion())) {
```
I think this can be better put in the findmaster() logic. If the "elected" master is not of a good version we should log a warning and return null (no master found)

martijnvg replied:
good point

bleskes commented on core/src/main/java/org/elasticsearch/Version.java:
```diff
@@ -584,6 +591,15 @@ public Version minimumCompatibilityVersion() {
+     * version. Beta and RC releases are always forward compatible with the maximum
+     * hypothetical version.
+     */
+    public Version maximumCompatibilityVersion() {
```
I think this is tricky (thought about it some more). We now say that 2.0 is incompatible with 1.x  but maybe in the future we will decide differently. For example, that 3.0 is compatible with 2.x. We will only make this decision when we come close to 3.0 so we can not "bake it" into 2.0 code now. Since we now add a min compatible check both on the node and master side during join - I don't think we need a max check?

martijnvg replied:
that question did cross my mind too, but historically this is what always has happened, so that was the main drive to add it... but yes since the minimum version is checked on both sides, checking the maximum version isn't necessary.

bleskes commented:
Left some comments..

martijnvg commented:
@bleskes thanks for the feedback! I updated the PR. Instead of moving the version check to findMaster(), I moved it to ElectMasterService#electMaster(), so the check doesn't need to get duplicated multiple times.

bleskes commented on core/src/main/java/org/elasticsearch/discovery/zen/ZenDiscovery.java:
```diff
@@ -886,12 +889,19 @@ static boolean shouldIgnoreOrRejectNewClusterState(ESLogger logger, ClusterState
             // TODO, what should we do now? Maybe inform that node that its crap?
             logger.warn("received a wrong address type from [{}], ignoring...", node);
         } else {
+            // Sanity checks: likely we don't end up here, because serialization is likely to fail.
```
why is serailization likely to fail? it might, it might now...

martijnvg replied:
maybe my choice of wording here... I'll change it to: `we may not end up here.....`

bleskes commented on core/src/main/java/org/elasticsearch/discovery/zen/ZenDiscovery.java:
```diff
@@ -886,12 +889,19 @@ static boolean shouldIgnoreOrRejectNewClusterState(ESLogger logger, ClusterState
         } else {
+            // Sanity checks: likely we don't end up here, because serialization is likely to fail.
+            if (node.getVersion().before(version.minimumCompatibilityVersion())) {
+                Exception e = new IllegalStateException("Can't handle join request from a node with a version [" + node.getVersion() + "] that is lower than the minimum compatible version [" + version.minimumCompatibilityVersion() + "]");
```
inline the variable in the onFailure?

martijnvg replied:
sure

bleskes commented on core/src/main/java/org/elasticsearch/discovery/zen/elect/ElectMasterService.java:
```diff
@@ -36,13 +37,15 @@
 
     @Inject
-    public ElectMasterService(Settings settings) {
+    public ElectMasterService(Settings settings, Version version) {
```
can we rename version into minMasterVersion, call it with Version. minimumCompatibilityVersion() and document what it means? I think it will be clearer.

martijnvg replied:
agreed

bleskes replied:
It turns out it's hard(ish) to remove ElectMasterService from guice.. not doing.

bleskes commented:
Thx @martijnvg . I left some minor comments.

martijnvg commented:
@bleskes cool, I've updated the PR.

martijnvg commented:
@bleskes I change `ElectMasterService` instance to be managed by `ZenDiscovery`. I'm not happy about it, because it used by `ZenPingService` (which is injected by `ZenDiscovery` too). I made this work by making `ZenPingService#sortByMasterLikelihood()` static.

martijnvg commented:
@bleskes I reverted the previous commit and let `ZenDiscovery` get the current version from the local node.

bleskes commented on core/src/main/java/org/elasticsearch/discovery/zen/ZenDiscovery.java:
```diff
@@ -886,12 +890,20 @@ static boolean shouldIgnoreOrRejectNewClusterState(ESLogger logger, ClusterState
     }
 
-    private void handleJoinRequest(final DiscoveryNode node, final MembershipAction.JoinCallback callback) {
+    void handleJoinRequest(final DiscoveryNode node, final MembershipAction.JoinCallback callback) {
```
this can go back to private now, no?

martijnvg replied:
true

martijnvg replied:
oops... no, because it is used in a test (ZenDiscoveryTests#testHandleNodeJoin_incompatibleMinVersion)

bleskes commented:
LGTM. Left a couple of trivial comments. No need for another review.
